socket: Reduce the waiting time for socket connect

chenglin130 opened the pull request:
It's a too long time to sleep 1s waiting for unix socket ready.
10ms may be sufficient.
Reduce waiting time to improve efficiency and performance.

Fixes: #2037

Signed-off-by: Cheng Lin <cheng.lin130@zte.com.cn>

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

chenglin130 commented:
- Before this patch

glusterd log：
```
[2021-03-17 03:06:10.411201] T [socket.c:3490:socket_connect] 0-quotad: >>> connect() with non-blocking IO for ALL
[2021-03-17 03:06:10.411217] T [socket.c:224:socket_dump_info] 0-quotad: $$$ client: connecting to (af:1,sock:80) /var/run/gluster/1e4e32f615e8d059.socket non-SSL (errno:0:Success)
```

quotad log： 
```
[2021-03-17 03:06:10.418365] T [socket.c:224:socket_dump_info] 0-socket.glusterfsd: $$$ client: listening on (af:1,sock:7) /var/run/gluster/1e4e32f615e8d059.socket non-SSL (errno:0:Success)
[2021-03-17 03:06:11.411775] T [socket.c:3105:socket_server_event_handler] 0-socket.glusterfsd: XXX server:/var/run/gluster/1e4e32f615e8d059.socket, client:
```

```
# time gluster volume quota gv1 enable
volume quota : success

real    0m2.199s
user    0m0.026s
sys     0m0.025s
```

- After this patch：

```
# time gluster volume quota gv1 enable
volume quota : success

real    0m0.223s
user    0m0.031s
sys     0m0.019s
```

mohit84 commented:
/run regression

mykaul commented:
Please see https://review.gluster.org/#/c/glusterfs/+/22632/ for a similar, perhaps more comprehensive effort. I do believe it's worthwhile to do, but we've had some failures I believe. @xhernandez may shed some more light on this.

chenglin130 commented:
> Please see https://review.gluster.org/#/c/glusterfs/+/22632/ for a similar, perhaps more comprehensive effort. I do believe it's worthwhile to do, but we've had some failures I believe. @xhernandez may shed some more light on this.

Great minds think alike. 

The logs show that 10 ms is proper than 1s（or 100ms）. 
 test for sleep 100ms 
```
# time gluster volume quota gv1 enable
volume quota : success

real    0m0.387s
user    0m0.024s
sys     0m0.021s
```

Compared with 100 ms,  10ms will doubles the speed.
Considerable benefits.

xhernandez commented:
I don't remember exactly what happened to the old patch. Besides it languishing for a long time under review, probably the higher frequency of checks triggered some unexpected issue during regression run. Regression tests have been improved since thena and this patch doesn't touch as many timeouts as the other, so maybe it has better chances to work. Let's seen the result of regression run.

Regarding the delay between 10 and 100 ms, I still think that 100 ms is better, because increasing 10 times the frequency (so using 10x more CPU) only provides less than 50% improvement (assuming that time results are reliable, which are not for so small delays). But I'm not agains using 10 ms if others agree.

xhernandez commented:
/run regression

chenglin130 commented:
> I don't remember exactly what happened to the old patch. Besides it languishing for a long time under review, probably the higher frequency of checks triggered some unexpected issue during regression run. Regression tests have been improved since thena and this patch doesn't touch as many timeouts as the other, so maybe it has better chances to work. Let's seen the result of regression run.
> 
> Regarding the delay between 10 and 100 ms, I still think that 100 ms is better, because increasing 10 times the frequency (so using 10x more CPU) only provides less than 50% improvement (assuming that time results are reliable, which are not for so small delays). But I'm not agains using 10 ms if others agree.

@xhernandez @mohit84 @mykaul 
In my test environment, quotad socket will be ready within 10ms. This may be the time used in most cases.
If the first delay time is 100ms, then more than 90% of the waiting time is unnecessary.

Alternatively, we can do some times fast check (e.g. 3times 10ms delay) firstly, and then do other slow check (100ms delay).
I believe that the fast check at the beginning will deal with the situation above 90%.

In this way, the maximum benefits can be obtained at the minimum price.

xhernandez commented:
/run regression

xhernandez commented:
The patch seems ok to me. I'll wait until regression finishes, and for other reviews.

chenglin130 commented:
> The patch seems ok to me. I'll wait until regression finishes, and for other reviews.
@xhernandez 

the failed testcase:
```
00:15:06 not ok  17 [     11/  45121] <  43> 'trusted.bit-rot.signature check_for_xattr trusted.bit-rot.signature //d/backends/patchy1/FILE1' -> 'Got "" instead of "trusted.bit-rot.signature"'
00:15:06 ok  18 [     18/      1] <  46> ''
00:15:06 ok  19 [     10/     81] <  49> 'gluster --mode=script --wignore volume bitrot patchy scrub ondemand'
00:15:06 not ok  20 [     11/  45223] <  50> 'trusted.bit-rot.bad-file check_for_xattr trusted.bit-rot.bad-file //d/backends/patchy1/FILE1' -> 'Got "" instead of "trusted.bit-rot.bad-file"'
00:15:06 Failed 2/20 subtests

00:15:06 ./tests/bitrot/bug-1207627-bitrot-scrub-status.t (Wstat: 0 Tests: 20 Failed: 2)
00:15:06   Failed tests:  17, 20

```
In my environment, the testcase passed.
```
# prove -vmfe '/bin/bash' tests/bitrot/bug-1207627-bitrot-scrub-status.t
tests/bitrot/bug-1207627-bitrot-scrub-status.t ..
1..20
ok   1 [    626/   1014] <  11> 'glusterd'
ok   2 [     96/     23] <  12> 'pidof glusterd'
ok   3 [    100/     68] <  15> 'gluster --mode=script --wignore volume create patchy glst-srv-1:/d/backends/patchy1'
ok   4 [     97/    257] <  16> 'gluster --mode=script --wignore volume start patchy'
ok   5 [    134/    108] <  19> 'gluster --mode=script --wignore volume bitrot patchy enable'
ok   6 [    139/     32] <  21> '1 get_bitd_count'
ok   7 [    105/     61] <  24> 'gluster --mode=script --wignore volume bitrot patchy scrub-frequency hourly'
ok   8 [     98/     59] <  27> 'gluster --mode=script --wignore volume bitrot patchy scrub-throttle lazy'
ok   9 [     97/     67] <  29> 'Active scrub_status patchy State of scrub'
ok  10 [    121/     98] <  30> 'lazy scrub_status patchy Scrub impact'
ok  11 [    102/     69] <  31> 'hourly scrub_status patchy Scrub frequency'
ok  12 [     99/     64] <  32> '/var/log/glusterfs/bitd.log scrub_status patchy Bitrot error log location'
ok  13 [     99/     64] <  33> '/var/log/glusterfs/scrub.log scrub_status patchy Scrubber error log location'
ok  14 [     99/     68] <  36> 'gluster --mode=script --wignore volume set patchy features.expiry-time 1'
ok  15 [     97/    246] <  39> '_GFS --attribute-timeout=0 --entry-timeout=0 --volfile-id=patchy --volfile-server=glst-srv-1 /mnt/glusterfs/0'
ok  16 [    157/      3] <  42> ''
ok  17 [    101/   1036] <  43> 'trusted.bit-rot.signature check_for_xattr trusted.bit-rot.signature //d/backends/patchy1/FILE1'
ok  18 [     98/      1] <  46> ''
ok  19 [    100/     66] <  49> 'gluster --mode=script --wignore volume bitrot patchy scrub ondemand'
ok  20 [    101/   1553] <  50> 'trusted.bit-rot.bad-file check_for_xattr trusted.bit-rot.bad-file //d/backends/patchy1/FILE1'
ok
All tests successful.
Files=1, Tests=20,  8 wallclock secs ( 0.02 usr  0.01 sys +  1.39 cusr  1.97 csys =  3.39 CPU)
Result: PASS
```

mohit84 commented:
/run regression

chenglin130 commented:
The same errors have been happened again. 
No found Expected xattrs on the tested file.

```
13:50:19 ======================================== (265 / 785) ========================================
13:50:19 [05:50:19] Running tests in file ./tests/bitrot/bug-1207627-bitrot-scrub-status.t
13:50:19 ./tests/bitrot/bug-1207627-bitrot-scrub-status.t .. 
13:51:53 1..20
13:51:53 ok   1 [    201/   1269] <  11> 'glusterd'
13:51:53 ok   2 [     10/     27] <  12> 'pidof glusterd'
13:51:53 ok   3 [      9/     78] <  15> 'gluster --mode=script --wignore volume create patchy builder202.int.aws.gluster.org:/d/backends/patchy1'
13:51:53 ok   4 [     12/    308] <  16> 'gluster --mode=script --wignore volume start patchy'
13:51:53 ok   5 [     13/    136] <  19> 'gluster --mode=script --wignore volume bitrot patchy enable'
13:51:53 ok   6 [     30/     28] <  21> '1 get_bitd_count'
13:51:53 ok   7 [     18/     94] <  24> 'gluster --mode=script --wignore volume bitrot patchy scrub-frequency hourly'
13:51:53 ok   8 [     12/     94] <  27> 'gluster --mode=script --wignore volume bitrot patchy scrub-throttle lazy'
13:51:53 ok   9 [     12/     99] <  29> 'Active scrub_status patchy State of scrub'
13:51:53 ok  10 [     14/     93] <  30> 'lazy scrub_status patchy Scrub impact'
13:51:53 ok  11 [     12/     94] <  31> 'hourly scrub_status patchy Scrub frequency'
13:51:53 ok  12 [     13/     93] <  32> '/var/log/glusterfs/bitd.log scrub_status patchy Bitrot error log location'
13:51:53 ok  13 [     13/     96] <  33> '/var/log/glusterfs/scrub.log scrub_status patchy Scrubber error log location'
13:51:53 ok  14 [     11/    103] <  36> 'gluster --mode=script --wignore volume set patchy features.expiry-time 1'
13:51:53 ok  15 [     12/     50] <  39> '_GFS --attribute-timeout=0 --entry-timeout=0 --volfile-id=patchy --volfile-server=builder202.int.aws.gluster.org /mnt/glusterfs/0'
13:51:53 ok  16 [     15/      1] <  42> ''
13:51:53 not ok  17 [     11/  45134] <  43> 'trusted.bit-rot.signature check_for_xattr trusted.bit-rot.signature //d/backends/patchy1/FILE1' -> 'Got "" instead of "trusted.bit-rot.signature"'
13:51:53 ok  18 [     14/      1] <  46> ''
13:51:53 ok  19 [     11/     84] <  49> 'gluster --mode=script --wignore volume bitrot patchy scrub ondemand'
13:51:53 not ok  20 [     12/  45170] <  50> 'trusted.bit-rot.bad-file check_for_xattr trusted.bit-rot.bad-file //d/backends/patchy1/FILE1' -> 'Got "" instead of "trusted.bit-rot.bad-file"'
13:51:53 Failed 2/20 subtests 
13:51:53 
13:51:53 Test Summary Report
13:51:53 -------------------
13:51:53 ./tests/bitrot/bug-1207627-bitrot-scrub-status.t (Wstat: 0 Tests: 20 Failed: 2)
13:51:53   Failed tests:  17, 20
13:51:53 Files=1, Tests=20, 94 wallclock secs ( 0.03 usr  0.00 sys +  2.85 cusr  2.23 csys =  5.11 CPU)
13:51:53 Result: FAIL
13:51:53 Logs preserved in tarball bug-1207627-bitrot-scrub-status-iteration-1.tar
13:51:53 ./tests/bitrot/bug-1207627-bitrot-scrub-status.t: bad status 1
```
@xhernandez @mohit84 @mykaul 
I found that there are a time delay between triggering operation and actually setting xattr.  May be need a little wait before check.
```
[root@glst-srv-1 test]# echo 2222 >> /data/brick2/gv-s-0/file4
[root@glst-srv-1 test]# getfattr -d -m. -e hex /data/brick2/gv-s-0/file4
getfattr: Removing leading '/' from absolute path names
# file: data/brick2/gv-s-0/file4
security.selinux=0x73797374656d5f753a6f626a6563745f723a676c7573746572645f627269636b5f743a733000
trusted.bit-rot.signature=0x010200000000000000f4dd4112a2540430e5a8a7158bcf120f8f2489bffa4d8f6aefa8dcd584047011
trusted.bit-rot.version=0x02000000000000006055b681000ebb16
trusted.gfid=0xd1d45ffa124d46dc825c279e63bf92c7
trusted.gfid2path.fdfa64a54da7c027=0x30303030303030302d303030302d303030302d303030302d3030303030303030303030312f66696c6534
trusted.glusterfs.mdata=0x010000000000000000000000006055b738000000001640ffd2000000006055b738000000001640ffd2000000006055b73800000000162a6af9

[root@glst-srv-1 test]# gluster volume bitrot gv-s-0 scrub ondemand
volume bitrot: scrubber started ondemand for volume gv-s-0
[root@glst-srv-1 test]# getfattr -d -m. -e hex /data/brick2/gv-s-0/file4
getfattr: Removing leading '/' from absolute path names
# file: data/brick2/gv-s-0/file4
security.selinux=0x73797374656d5f753a6f626a6563745f723a676c7573746572645f627269636b5f743a733000
trusted.bit-rot.signature=0x010200000000000000f4dd4112a2540430e5a8a7158bcf120f8f2489bffa4d8f6aefa8dcd584047011
trusted.bit-rot.version=0x02000000000000006055b681000ebb16
trusted.gfid=0xd1d45ffa124d46dc825c279e63bf92c7
trusted.gfid2path.fdfa64a54da7c027=0x30303030303030302d303030302d303030302d303030302d3030303030303030303030312f66696c6534
trusted.glusterfs.mdata=0x010000000000000000000000006055b738000000001640ffd2000000006055b738000000001640ffd2000000006055b73800000000162a6af9

[root@glst-srv-1 test]# getfattr -d -m. -e hex /data/brick2/gv-s-0/file4
getfattr: Removing leading '/' from absolute path names
# file: data/brick2/gv-s-0/file4
security.selinux=0x73797374656d5f753a6f626a6563745f723a676c7573746572645f627269636b5f743a733000
trusted.bit-rot.bad-file=0x3100
trusted.bit-rot.signature=0x010200000000000000f4dd4112a2540430e5a8a7158bcf120f8f2489bffa4d8f6aefa8dcd584047011
trusted.bit-rot.version=0x02000000000000006055b681000ebb16
trusted.gfid=0xd1d45ffa124d46dc825c279e63bf92c7
trusted.gfid2path.fdfa64a54da7c027=0x30303030303030302d303030302d303030302d303030302d3030303030303030303030312f66696c6534
trusted.glusterfs.mdata=0x010000000000000000000000006055b738000000001640ffd2000000006055b738000000001640ffd2000000006055b73800000000162a6af9

[root@glst-srv-1 test]#
```

chenglin130 commented:
..

xhernandez commented:
> The same errors have been happened again.
> No found Expected xattrs on the tested file.
> 
> [...]
> 
> @xhernandez @mohit84 @mykaul
> I found that there are a time delay between triggering operation and actually setting xattr. May be need a little wait before check.
> 

How is this related to this change ? this change only makes the process to complete faster, but how does it alter the delay of updating xattrs after a change ?

chenglin130 commented:
> > The same errors have been happened again.
> > No found Expected xattrs on the tested file.
> > [...]
> > @xhernandez @mohit84 @mykaul
> > I found that there are a time delay between triggering operation and actually setting xattr. May be need a little wait before check.
> 
> How is this related to this change ? this change only makes the process to complete faster, but how does it alter the delay of updating xattrs after a change ?

In my test enviroment , a 4 processers virtual machine, the time delay is similar before and after this change. And the testcase is PASSed. It seems that this change will not lead to the error.
In the error, check_for_xattr failed with a timeout 45s (No need wait, EXPECT_WITHIN will do it. I didn't realize it before.).   
`13:51:53 not ok  17 [     11/  45134] <  43> 'trusted.bit-rot.signature check_for_xattr trusted.bit-rot.signature //d/backends/patchy1/FILE1' -> 'Got "" instead of "trusted.bit-rot.signature"'`

I'm puzzled too now. 
Does bug-1207627-bit rot-scrub-status-iteration-1.tar have enough information to analyze the error?

chenglin130 commented:
@xhernandez 
In glusterd.log,  every op follows a “Dict fialed”.

```
[2021-03-18 05:50:21.212698]:++++++++++ G_LOG:./tests/bitrot/bug-1207627-bitrot-scrub-status.t: TEST: 15 gluster --mode=script --wignore volume create patchy builder202.int.aws.gluster.org:/d/backends/patchy1 ++++++++++
[2021-03-18 05:50:21.268182 +0000] E [MSGID: 101032] [store.c:464:gf_store_handle_retrieve] 0-: Path corresponding to /var/lib/glusterd/glusterd.info. [No such file or directory]
[2021-03-18 05:50:21.268226 +0000] I [MSGID: 106477] [glusterd.c:181:glusterd_uuid_generate_save] 0-management: generated UUID: b614437e-01c0-4798-b645-b12c9fe1194e 
[2021-03-18 05:50:21.272339 +0000] E [MSGID: 106048] [glusterd-server-quorum.c:100:glusterd_validate_quorum] 0-management: Failed to get volinfo [] 
[2021-03-18 05:50:21.286981 +0000] E [MSGID: 106061] [glusterd-volgen.c:2520:brick_graph_add_pump] 0-management: Dict get failed [{Key=enable-pump}, {errno=2}, {error=No such file or directory}] 
[2021-03-18 05:50:21.302456]:++++++++++ G_LOG:./tests/bitrot/bug-1207627-bitrot-scrub-status.t: TEST: 16 gluster --mode=script --wignore volume start patchy ++++++++++
[2021-03-18 05:50:21.375705 +0000] E [MSGID: 106061] [glusterd-server-quorum.c:258:glusterd_is_volume_in_server_quorum] 0-management: Dict get failed [{Key=cluster.server-quorum-type}] 
...
[2021-03-18 05:50:21.838808]:++++++++++ G_LOG:./tests/bitrot/bug-1207627-bitrot-scrub-status.t: TEST: 24 gluster --mode=script --wignore volume bitrot patchy scrub-frequency hourly ++++++++++
[2021-03-18 05:50:21.908861 +0000] E [MSGID: 106061] [glusterd-server-quorum.c:258:glusterd_is_volume_in_server_quorum] 0-management: Dict get failed [{Key=cluster.server-quorum-type}] 
[2021-03-18 05:50:21.913630 +0000] E [MSGID: 106061] [glusterd-volgen.c:2520:brick_graph_add_pump] 0-management: Dict get failed [{Key=enable-pump}, {errno=2}, {error=No such file or directory}] 
...
[2021-03-18 05:51:07.949776]:++++++++++ G_LOG:./tests/bitrot/bug-1207627-bitrot-scrub-status.t: TEST: 49 gluster --mode=script --wignore volume bitrot patchy scrub ondemand ++++++++++
[2021-03-18 05:51:08.019473 +0000] E [MSGID: 106061] [glusterd-server-quorum.c:258:glusterd_is_volume_in_server_quorum] 0-management: Dict get failed [{Key=cluster.server-quorum-type}] 
[2021-03-18 05:51:08.020382 +0000] E [MSGID: 106061] [glusterd-volgen.c:2520:brick_graph_add_pump] 0-management: Dict get failed [{Key=enable-pump}, {errno=2}, {error=No such file or directory}] 
```

In mnt-glusterfs-0.log, found fuse warn 'Operation not permitted'.

```
[2021-03-18 05:50:22.739856 +0000] I [MSGID: 114057] [client-handshake.c:1128:select_server_supported_programs] 0-patchy-client-0: Using Program [{Program-name=GlusterFS 4.x v1}, {Num=1298437}, {Version=400}] 
[2021-03-18 05:50:22.740535 +0000] I [MSGID: 114046] [client-handshake.c:857:client_setvolume_cbk] 0-patchy-client-0: Connected, attached to remote volume [{conn-name=patchy-client-0}, {remote_subvol=/d/backends/patchy1}] 
[2021-03-18 05:50:22.742189 +0000] W [fuse-bridge.c:6182:fuse_thread_proc] 0-glusterfs-fuse: read from /dev/fuse returned -1 (Operation not permitted)
[2021-03-18 05:50:22.742282 +0000] I [MSGID: 101190] [event-epoll.c:670:event_dispatch_epoll_worker] 0-epoll: Started thread with index [{index=1}] 
[2021-03-18 05:50:22.753543 +0000] I [fuse-bridge.c:5295:fuse_init] 0-glusterfs-fuse: FUSE inited with protocol versions: glusterfs 7.24 kernel 7.23
[2021-03-18 05:50:22.753573 +0000] I [fuse-bridge.c:5923:fuse_graph_sync] 0-fuse: switched to graph 0
```

In bitd.log, found 'signing' record.

```
[2021-03-18 05:50:22.774648]:++++++++++ G_LOG:./tests/bitrot/bug-1207627-bitrot-scrub-status.t: TEST: 42 ++++++++++
[2021-03-18 05:50:22.787302]:++++++++++ G_LOG:./tests/bitrot/bug-1207627-bitrot-scrub-status.t: TEST: 43 trusted.bit-rot.signature check_for_xattr trusted.bit-rot.signature //d/backends/patchy1/FILE1 ++++++++++
[2021-03-18 05:50:23.795181 +0000] I [MSGID: 118039] [bit-rot.c:1268:br_child_enaction] 0-patchy-bit-rot-0: Connected to brick [{brick-path=/d/backends/patchy1}] 
[2021-03-18 05:50:23.795257 +0000] I [MSGID: 118018] [bit-rot.c:1088:br_oneshot_signer] 0-patchy-bit-rot-0: Crawling brick, scanning for unsigned objects [{brick-path=/d/backends/patchy1}] 
[2021-03-18 05:50:23.796691 +0000] I [MSGID: 118016] [bit-rot.c:1054:bitd_oneshot_crawl] 0-patchy-bit-rot-0: Triggering signing [{path=/FILE1}, {gfid=3a516625-02fe-4afb-9cde-25fb379ecbd6}, {Brick-path=/d/backends/patchy1}] 
[2021-03-18 05:50:23.797351 +0000] I [MSGID: 118025] [bit-rot.c:1096:br_oneshot_signer] 0-patchy-bit-rot-0: Completed crawling brick [{brick-path=/d/backends/patchy1}] 
[2021-03-18 05:51:07.937164]:++++++++++ G_LOG:./tests/bitrot/bug-1207627-bitrot-scrub-status.t: TEST: 46 ++++++++++
[2021-03-18 05:51:07.949776]:++++++++++ G_LOG:./tests/bitrot/bug-1207627-bitrot-scrub-status.t: TEST: 49 gluster --mode=script --wignore volume bitrot patchy scrub ondemand ++++++++++
```

In scrub.log, no found 'CORRUPTION DETECTED' record, just 'Scrubbing started' 'Scrubbing started'.

```
[2021-03-18 05:51:07.949776]:++++++++++ G_LOG:./tests/bitrot/bug-1207627-bitrot-scrub-status.t: TEST: 49 gluster --mode=script --wignore volume bitrot patchy scrub ondemand ++++++++++
[2021-03-18 05:51:08.019689 +0000] I [bit-rot.c:1690:notify] 0-patchy-bit-rot-0: BitRot scrub ondemand called
[2021-03-18 05:51:08.019711 +0000] I [MSGID: 118038] [bit-rot-scrub.c:1093:br_fsscan_ondemand] 0-patchy-bit-rot-0: Ondemand Scrubbing scheduled to run at 2021-03-18 05:51:09 +0000 
[2021-03-18 05:51:08.026918 +0000] I [glusterfsd-mgmt.c:77:mgmt_cbk_spec] 0-mgmt: Volume file changed
[2021-03-18 05:51:08.035886 +0000] I [glusterfsd-mgmt.c:2238:mgmt_getspec_cbk] 0-glusterfs: No change in volfile,continuing
[2021-03-18 05:51:08.046014]:++++++++++ G_LOG:./tests/bitrot/bug-1207627-bitrot-scrub-status.t: TEST: 50 trusted.bit-rot.bad-file check_for_xattr trusted.bit-rot.bad-file //d/backends/patchy1/FILE1 ++++++++++
[2021-03-18 05:51:09.789752 +0000] I [MSGID: 118044] [bit-rot-scrub.c:618:br_scrubber_log_time] 0-patchy-bit-rot-0: Scrubbing started at 2021-03-18 05:51:09 +0000 
[2021-03-18 05:51:09.792100 +0000] I [MSGID: 118045] [bit-rot-scrub.c:622:br_scrubber_log_time] 0-patchy-bit-rot-0: Scrubbing finished at 2021-03-18 05:51:09 +0000 
[2021-03-18 05:51:09.792129 +0000] I [MSGID: 118038] [bit-rot-scrub.c:1003:br_fsscan_activate] 0-patchy-bit-rot-0: Scrubbing is rescheduled to run at 2021-03-18 06:51:09 +0000 
[2021-03-18 05:51:09.792141 +0000] I [MSGID: 118038] [bit-rot-scrub.c:1905:wait_for_scrub_to_finish] 0-patchy-bit-rot-0: Waiting for all children to start and finish scrub 
```

like this (in no failed test log):
```
[2021-03-23 10:46:03.997403] A [MSGID: 118023] [bit-rot-scrub.c:241:bitd_compare_ckum] 0-patchy-bit-rot-0: CORRUPTION DETECTED: Object /FILE1 {Brick: /d/backends/patchy1 | GFID: 8320db17-1f02-4261-9c7e-4585547e3eab}
[2021-03-23 10:46:03.997457] A [MSGID: 118024] [bit-rot-scrub.c:262:bitd_compare_ckum] 0-patchy-bit-rot-0: Marking /FILE1 [GFID: 8320db17-1f02-4261-9c7e-4585547e3eab | Brick: /d/backends/patchy1] as corrupted..
```

chenglin130 commented:
could run regression again？
let me collect another log.

chenglin130 commented:
This time "tests/bitrot/bug-1207627-bitrot-scrub-status.t" is PASSed , but "tests/bitrot/bug-1700078.t" is failed.
the reason is the same one, failed to get file's xattr.

```
12:40:15 ======================================== (270 / 786) ========================================
12:40:15 [04:40:15] Running tests in file ./tests/bitrot/bug-1700078.t
12:41:09 ./tests/bitrot/bug-1700078.t .. 
12:41:09 1..27
12:41:09 ok   1 [    212/   1414] <   9> 'glusterd'
12:41:09 ok   2 [     11/     16] <  10> 'pidof glusterd'
12:41:09 ok   3 [     12/     96] <  13> 'gluster --mode=script --wignore volume create patchy builder202.int.aws.gluster.org:/d/backends/patchy1'
12:41:09 ok   4 [    161/    515] <  14> 'gluster --mode=script --wignore volume start patchy'
12:41:09 ok   5 [     15/    101] <  17> 'gluster --mode=script --wignore volume bitrot patchy enable'
12:41:09 ok   6 [     13/     77] <  23> 'gluster --mode=script --wignore volume set patchy performance.quick-read off'
12:41:09 ok   7 [      9/     29] <  25> '1 get_bitd_count'
12:41:09 ok   8 [      9/     73] <  27> 'Active scrub_status patchy State of scrub'
12:41:09 ok   9 [      9/     77] <  28> '/var/log/glusterfs/bitd.log scrub_status patchy Bitrot error log location'
12:41:09 ok  10 [      9/     71] <  29> '/var/log/glusterfs/scrub.log scrub_status patchy Scrubber error log location'
12:41:09 ok  11 [      9/     81] <  32> 'gluster --mode=script --wignore volume set patchy features.expiry-time 1'
12:41:09 ok  12 [     10/     21] <  35> 'glusterfs --volfile-id=patchy --volfile-server=builder202.int.aws.gluster.org /mnt/glusterfs/0 --attribute-timeout=0 --entry-timeout=0'
12:41:09 ok  13 [     11/     78] <  39> 'gluster --mode=script --wignore volume set patchy performance.quick-read off'
12:41:09 ok  14 [     14/      1] <  42> ''
12:41:09 not ok  15 [     10/  45234] <  43> 'trusted.bit-rot.signature check_for_xattr trusted.bit-rot.signature //d/backends/patchy1/FILE1' -> 'Got "" instead of "trusted.bit-rot.signature"'
12:41:09 ok  16 [     12/     80] <  46> 'gluster --mode=script --wignore volume bitrot patchy disable'
12:41:09 ok  17 [     15/      1] <  49> ''
12:41:09 ok  18 [     11/     17] <  55> 'Y force_umount /mnt/glusterfs/0'
12:41:09 ok  19 [      9/    106] <  58> 'gluster --mode=script --wignore volume bitrot patchy enable'
12:41:09 ok  20 [   2016/     40] <  64> '1 get_bitd_count'
12:41:09 ok  21 [     12/     85] <  66> 'Active scrub_status patchy State of scrub'
12:41:09 ok  22 [     11/     85] <  67> '/var/log/glusterfs/bitd.log scrub_status patchy Bitrot error log location'
12:41:09 ok  23 [     11/     87] <  68> '/var/log/glusterfs/scrub.log scrub_status patchy Scrubber error log location'
12:41:09 ok  24 [     11/     95] <  71> 'gluster --mode=script --wignore volume bitrot patchy scrub ondemand'
12:41:09 ok  25 [   2012/      3] <  80> '! getfattr -n trusted.bit-rot.bad-file /d/backends/patchy1/FILE1'
12:41:09 ok  26 [     12/     20] <  83> 'glusterfs --volfile-id=patchy --volfile-server=builder202.int.aws.gluster.org /mnt/glusterfs/0 --attribute-timeout=0 --entry-timeout=0'
12:41:09 ok  27 [     12/      4] <  85> 'cat /mnt/glusterfs/0/FILE1'
12:41:09 Failed 1/27 subtests 
12:41:09 
12:41:09 Test Summary Report
12:41:09 -------------------
12:41:09 ./tests/bitrot/bug-1700078.t (Wstat: 0 Tests: 27 Failed: 1)
12:41:09   Failed test:  15
12:41:09 Files=1, Tests=27, 54 wallclock secs ( 0.02 usr  0.00 sys +  2.08 cusr  1.43 csys =  3.53 CPU)
12:41:09 Result: FAIL
12:41:09 Logs preserved in tarball bug-1700078-iteration-1.tar
12:41:09 ./tests/bitrot/bug-1700078.t: bad status 1
```

in bitd.log, can found record "Triggering signing [{path=/FILE1}", and no errors report.

```
[2021-03-30 04:41:12.259205]:++++++++++ G_LOG:./tests/bitrot/bug-1700078.t: TEST: 42 ++++++++++
[2021-03-30 04:41:12.274284]:++++++++++ G_LOG:./tests/bitrot/bug-1700078.t: TEST: 43 trusted.bit-rot.signature check_for_xattr trusted.bit-rot.signature //d/backends/patchy1/FILE1 ++++++++++
[2021-03-30 04:41:13.458192 +0000] I [MSGID: 118039] [bit-rot.c:1268:br_child_enaction] 0-patchy-bit-rot-0: Connected to brick [{brick-path=/d/backends/patchy1}] 
[2021-03-30 04:41:13.458275 +0000] I [MSGID: 118018] [bit-rot.c:1088:br_oneshot_signer] 0-patchy-bit-rot-0: Crawling brick, scanning for unsigned objects [{brick-path=/d/backends/patchy1}] 
[2021-03-30 04:41:13.459867 +0000] I [MSGID: 118016] [bit-rot.c:1054:bitd_oneshot_crawl] 0-patchy-bit-rot-0: Triggering signing [{path=/FILE1}, {gfid=1998f84d-5513-47f2-a2bc-bf4e2b5500d4}, {Brick-path=/d/backends/patchy1}] 
[2021-03-30 04:41:13.460508 +0000] I [MSGID: 118025] [bit-rot.c:1096:br_oneshot_signer] 0-patchy-bit-rot-0: Completed crawling brick [{brick-path=/d/backends/patchy1}] 
[2021-03-30 04:41:57.462082]:++++++++++ G_LOG:./tests/bitrot/bug-1700078.t: TEST: 46 gluster --mode=script --wignore volume bitrot patchy disable ++++++++++
```

And another testcase "tests/bitrot/bug-1373520.t", the first perform failed with "No such file or directory", and then try again is a PASSed.

```
12:39:59 stat: cannot stat ‘/d/backends/patchy5/HL_FILE1’: No such file or directory
12:39:59 stat: cannot stat ‘/d/backends/patchy5/HL_FILE1’: No such file or directory
12:39:59 stat: cannot stat ‘/d/backends/patchy5/HL_FILE1’: No such file or directory
12:39:59 not ok  31 [     12/  80176] <  69> '512 path_size /d/backends/patchy5/HL_FILE1' -> 'Got "" instead of "512"'
12:39:59 Failed 1/31 subtests 
12:39:59 
12:39:59 Test Summary Report
12:39:59 -------------------
12:39:59 ./tests/bitrot/bug-1373520.t (Wstat: 0 Tests: 31 Failed: 1)
12:39:59   Failed test:  31
12:39:59 Files=1, Tests=31, 95 wallclock secs ( 0.10 usr  0.00 sys +  3.04 cusr  2.28 csys =  5.42 CPU)
12:39:59 Result: FAIL
12:39:59 Logs preserved in tarball bug-1373520-iteration-1.tar
12:39:59 ./tests/bitrot/bug-1373520.t: bad status 1
```

The same file opration (in the same or different testcases), sometimes ok, sometimes failed.
Maybe there is something wrong with the backend storage filesystem?

chenglin130 commented:
Some debugging information is added. 
please run regression again.

xhernandez commented:
/run regression

chenglin130 commented:
> /run regression

please run regression again.

xhernandez commented:
/run regression

chenglin130 commented:
@xhernandez 
It seems, sometimes changelog failed to dispatch event to bitd.
why?

```
--- success ---
bitd log：
[2021-04-08 10:25:13.284126]:++++++++++ G_LOG:./tests/bugs/bitrot/bug-1228680.t: TEST: 36 _GFS --attribute-timeout=0 --entry-timeout=0 --volfile-server=builder202.int.aws.gluster.org --volfile-id=patchy /mnt/glusterfs/0 ++++++++++
[2021-04-08 10:25:13.983892 +0000] W [bit-rot.c:835:br_brick_callback] 0-patchy-bit-rot-0: XXX: got br_brick_callback [1] on da609608-952b-4de6-8e36-ad05a0f25afc
[2021-04-08 10:25:14.038902 +0000] I [MSGID: 118039] [bit-rot.c:1270:br_child_enaction] 0-patchy-bit-rot-0: Connected to brick [{brick-path=/d/backends/patchy1}] 
[2021-04-08 10:25:14.040834 +0000] I [MSGID: 118018] [bit-rot.c:1090:br_oneshot_signer] 0-patchy-bit-rot-0: Crawling brick, scanning for unsigned objects [{brick-path=/d/backends/patchy1}] 
[2021-04-08 10:25:14.041695 +0000] I [MSGID: 118025] [bit-rot.c:1098:br_oneshot_signer] 0-patchy-bit-rot-0: Completed crawling brick [{brick-path=/d/backends/patchy1}] 
[2021-04-08 10:25:17.988508 +0000] W [bit-rot.c:835:br_brick_callback] 0-patchy-bit-rot-0: XXX: got br_brick_callback [2] on da609608-952b-4de6-8e36-ad05a0f25afc
[2021-04-08 10:25:17.989190 +0000] W [bit-rot.c:610:br_sign_object] 0-patchy-bit-rot-0: XXX: Signing object [da609608-952b-4de6-8e36-ad05a0f25afc]
[2021-04-08 10:25:19.387423]:++++++++++ G_LOG:./tests/bugs/bitrot/bug-1228680.t: TEST: 46 getfattr -m . -n trusted.bit-rot.signature /d/backends/patchy0/filezero ++++++++++

brick log：
[2021-04-08 10:25:13.284126]:++++++++++ G_LOG:./tests/bugs/bitrot/bug-1228680.t: TEST: 36 _GFS --attribute-timeout=0 --entry-timeout=0 --volfile-server=builder202.int.aws.gluster.org --volfile-id=patchy /mnt/glusterfs/0 ++++++++++
[2021-04-08 10:25:13.331013 +0000] I [addr.c:54:compare_addr_and_update] 0-/d/backends/patchy0: allowed = "*", received addr = "172.31.31.76"
[2021-04-08 10:25:13.331037 +0000] I [login.c:110:gf_auth] 0-auth/login: allowed user names: 01860920-674a-4d64-9d6f-0fe3fae875ad
[2021-04-08 10:25:13.331050 +0000] I [MSGID: 115029] [server-handshake.c:566:server_setvolume] 0-patchy-server: accepted client from CTX_ID:fa5369b4-7475-4883-9dc6-00c4be3c9e62-GRAPH_ID:0-PID:6544-HOST:builder202.int.aws.gluster.org-PC_NAME:patchy-client-0-RECON_NO:-0 (version: 2021.04.08) with subvol /d/backends/patchy0 
[2021-04-08 10:25:13.351787 +0000] W [bit-rot-stub.c:3452:br_stub_release] 0-patchy-bitrot-stub: XXX: stub_release ret:0 [releaseversion: 0 | flags: 0 | signinfo: 0]
[2021-04-08 10:25:13.362737 +0000] W [posix-helpers.c:1306:posix_fhandle_pair] 0-patchy-posix: XXX: fsetxattr key:trusted.bit-rot.version for fd=275. (0)
[2021-04-08 10:25:13.363231 +0000] W [bit-rot-stub.c:3452:br_stub_release] 0-patchy-bitrot-stub: XXX: stub_release ret:1 [releaseversion: 2 | flags: 0 | signinfo: 1]
[2021-04-08 10:25:17.940398 +0000] W [bit-rot-stub.c:1417:br_stub_fsetxattr] 0-patchy-bitrot-stub: XXX: got br_stub_fsetxattr on da609608-952b-4de6-8e36-ad05a0f25afc
[2021-04-08 10:25:17.940430 +0000] W [bit-rot-stub.c:1450:br_stub_fsetxattr] 0-patchy-bitrot-stub: XXX: got REOPEN_SIGN_HINT on da609608-952b-4de6-8e36-ad05a0f25afc
[2021-04-08 10:25:17.940562 +0000] W [bit-rot-stub.c:3452:br_stub_release] 0-patchy-bitrot-stub: XXX: stub_release ret:1 [releaseversion: 2 | flags: 0 | signinfo: 2]
[2021-04-08 10:25:17.989750 +0000] W [bit-rot-stub.c:1417:br_stub_fsetxattr] 0-patchy-bitrot-stub: XXX: got br_stub_fsetxattr on da609608-952b-4de6-8e36-ad05a0f25afc
[2021-04-08 10:25:17.989780 +0000] W [bit-rot-stub.c:1420:br_stub_fsetxattr] 0-patchy-bitrot-stub: XXX: got SIGNATURE request on da609608-952b-4de6-8e36-ad05a0f25afc
[2021-04-08 10:25:17.989888 +0000] W [posix-helpers.c:1306:posix_fhandle_pair] 0-patchy-posix: XXX: fsetxattr key:trusted.bit-rot.signature for fd=275. (0)
[2021-04-08 10:25:19.387423]:++++++++++ G_LOG:./tests/bugs/bitrot/bug-1228680.t: TEST: 46 getfattr -m . -n trusted.bit-rot.signature /d/backends/patchy0/filezero ++++++++++


--- failed ---
bitd log：
[2021-04-08 10:25:01.595349]:++++++++++ G_LOG:./tests/bugs/bitrot/bug-1228680.t: TEST: 36 _GFS --attribute-timeout=0 --entry-timeout=0 --volfile-server=builder202.int.aws.gluster.org --volfile-id=patchy /mnt/glusterfs/0 ++++++++++
[2021-04-08 10:25:02.336369 +0000] I [MSGID: 118039] [bit-rot.c:1270:br_child_enaction] 0-patchy-bit-rot-0: Connected to brick [{brick-path=/d/backends/patchy0}] 
[2021-04-08 10:25:02.336446 +0000] I [MSGID: 118018] [bit-rot.c:1090:br_oneshot_signer] 0-patchy-bit-rot-0: Crawling brick, scanning for unsigned objects [{brick-path=/d/backends/patchy0}] 
[2021-04-08 10:25:02.337854 +0000] I [MSGID: 118016] [bit-rot.c:1056:bitd_oneshot_crawl] 0-patchy-bit-rot-0: Triggering signing [{path=/filezero}, {gfid=b5a76b27-e931-4c67-a0da-4e7d1b079019}, {Brick-path=/d/backends/patchy0}] 
[2021-04-08 10:25:02.338568 +0000] I [MSGID: 118025] [bit-rot.c:1098:br_oneshot_signer] 0-patchy-bit-rot-0: Completed crawling brick [{brick-path=/d/backends/patchy0}] 
[2021-04-08 10:25:07.695478]:++++++++++ G_LOG:./tests/bugs/bitrot/bug-1228680.t: TEST: 46 getfattr -m . -n trusted.bit-rot.signature /d/backends/patchy0/filezero ++++++++++

brick log:
[2021-04-08 10:25:01.595349]:++++++++++ G_LOG:./tests/bugs/bitrot/bug-1228680.t: TEST: 36 _GFS --attribute-timeout=0 --entry-timeout=0 --volfile-server=builder202.int.aws.gluster.org --volfile-id=patchy /mnt/glusterfs/0 ++++++++++
[2021-04-08 10:25:01.650202 +0000] I [addr.c:54:compare_addr_and_update] 0-/d/backends/patchy0: allowed = "*", received addr = "172.31.31.76"
[2021-04-08 10:25:01.650226 +0000] I [login.c:110:gf_auth] 0-auth/login: allowed user names: 9aad8c6c-5f37-4e75-8a0b-f7ff8910950d
[2021-04-08 10:25:01.650238 +0000] I [MSGID: 115029] [server-handshake.c:566:server_setvolume] 0-patchy-server: accepted client from CTX_ID:1e3e41d3-4791-4215-8e31-68b507b27ed8-GRAPH_ID:0-PID:6017-HOST:builder202.int.aws.gluster.org-PC_NAME:patchy-client-0-RECON_NO:-0 (version: 2021.04.08) with subvol /d/backends/patchy0 
[2021-04-08 10:25:01.657900 +0000] I [rpcsvc.c:852:rpcsvc_handle_rpc_call] 0-rpc-service: spawned a request handler thread for queue 1
[2021-04-08 10:25:01.665099 +0000] W [bit-rot-stub.c:3452:br_stub_release] 0-patchy-bitrot-stub: XXX: stub_release ret:0 [releaseversion: 0 | flags: 0 | signinfo: 0]
[2021-04-08 10:25:01.676713 +0000] W [posix-helpers.c:1306:posix_fhandle_pair] 0-patchy-posix: XXX: fsetxattr key:trusted.bit-rot.version for fd=274. (0)
[2021-04-08 10:25:01.677257 +0000] W [bit-rot-stub.c:3452:br_stub_release] 0-patchy-bitrot-stub: XXX: stub_release ret:1 [releaseversion: 2 | flags: 0 | signinfo: 1]
[2021-04-08 10:25:02.338178 +0000] W [bit-rot-stub.c:1417:br_stub_fsetxattr] 0-patchy-bitrot-stub: XXX: got br_stub_fsetxattr on b5a76b27-e931-4c67-a0da-4e7d1b079019
[2021-04-08 10:25:02.338199 +0000] W [bit-rot-stub.c:1450:br_stub_fsetxattr] 0-patchy-bitrot-stub: XXX: got REOPEN_SIGN_HINT on b5a76b27-e931-4c67-a0da-4e7d1b079019
[2021-04-08 10:25:02.338356 +0000] W [bit-rot-stub.c:3452:br_stub_release] 0-patchy-bitrot-stub: XXX: stub_release ret:1 [releaseversion: 2 | flags: 0 | signinfo: 2]
[2021-04-08 10:25:07.695478]:++++++++++ G_LOG:./tests/bugs/bitrot/bug-1228680.t: TEST: 46 getfattr -m . -n trusted.bit-rot.signature /d/backends/patchy0/filezero ++++++++++
```

More failed samples are needed, please run regression again.

xhernandez commented:
/run regression

chenglin130 commented:
The failure occurs before any testcase.
@xhernandez

xhernandez commented:
> The failure occurs before any testcase.
> @xhernandez

Hi @chenglin130. I haven't had time to look into this in detail, sorry. I'll try to run some tests to see if I can identify something, but I'm not sure when I'll be able to do that... I'll let you know when I have some data.

chenglin130 committed:
socket: Reduce the waiting time for socket connect

It's a too long time to sleep 1s waiting for unix socket ready.
Reduce waiting time to improve efficiency and performance.
Firstly, do 3 times fast check (delay 10ms); after then, do slow
check (delay 100ms).

Fixes: #2037

Signed-off-by: Cheng Lin <cheng.lin130@zte.com.cn>

xhernandez commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/bitrot/bug-1207627-bitrot-scrub-status.t

0 test(s) generated core 


5 test(s) needed retry 
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
./tests/basic/quick-read-with-upcall.t
./tests/basic/volume-snapshot.t
./tests/bitrot/bug-1207627-bitrot-scrub-status.t

2 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
https://build.gluster.org/job/gh_centos7-regression/1856/

chenglin130 commented:
It appears again.

xhernandez commented:
@kotreshhr could you take a look to see why bitrot is failing here ? I don't see anything obvious...
